include/oneapi/tbb/collaborative_call_once.h

```diff
@@ -0,0 +1,185 @@
+/*
+    Copyright (c) 2021 Intel Corporation
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#ifndef __TBB_collaborative_call_once_H
+#define __TBB_collaborative_call_once_H
+
+#if __TBB_PREVIEW_COLLABORATIVE_CALL_ONCE
+
+#include "tbb/task_arena.h"
+#include "tbb/task_group.h"
+#include "tbb/task.h"
+
+#include <atomic>
+
+namespace tbb {
+namespace detail {
+namespace d1 {
+
+#if _MSC_VER && !defined(__INTEL_COMPILER)
+    // Suppress warning: structure was padded due to alignment specifier
+    #pragma warning (push)
+    #pragma warning (disable: 4324)
+#endif
+
+constexpr std::size_t bit_count = 7;
+
+static_assert(1 << bit_count == max_nfs_size, "bit_count must be a log2(max_nfs_size)");
+
+inline constexpr std::uintptr_t maskoff_pointer(std::uintptr_t ptr) {
+    return ptr & (std::size_t(-1) << bit_count);
+}
+
+class alignas(max_nfs_size) run_once {
+    task_arena my_arena;
+    wait_context my_wait_context;
+    task_group_context my_context;
+
+    std::atomic<std::int64_t> ref_count{0};
+
+    template <typename Func>
+    class call_once_delegate : public delegate_base {
+    public:
+        call_once_delegate(Func& f) : my_func(f) {}
+
+        bool operator()() const override {
+            my_func();
+            return true;
+        }
+
+    private:
+        Func& my_func;
+    };
+
+    template<typename Fn>
+    void isolated_execute(Fn f) {
+        call_once_delegate<Fn> delegate(f);
+
+        r1::isolate_within_arena(delegate, reinterpret_cast<std::intptr_t>(this));
+    }
+
+public:
+    run_once()
+        : my_arena(task_arena::attach{})
+        , my_wait_context(0)
+        , my_context(task_group_context::bound,
+            task_group_context::default_traits | task_group_context::concurrent_wait)
+        {}
+
+    ~run_once() {
+        spin_wait_while(ref_count, [&](std::int64_t value) { return value > 0; }, std::memory_order_acquire);
+    }
+
+    void increase_ref() { ref_count++; }
+
+    void decrease_ref() { ref_count--; }
```

Comment:
As far as I know, `std::shared_lock` is available since C++14 standard. Minimal supported version of standard for oneTBB is C++11.

Reply:
Gotcha. Could use `lock()`/`unlock()` the same way; I'm not sure how others would feel about if that abuses convention.

Reply:
I like the idea but we limited with C++11. I am not sure that it makes sense to implement [Mutex named requirement](https://en.cppreference.com/w/cpp/named_req/Mutex) because the following code will be misleading that it is executed by multiple threads:
```
std::lock_guard lock(runner);
runner->wait();
```

Maybe it makes sense to use `try_call/on_copmletion` guard but see below. 

> runner->increaese_ref();
> ... // Be sure this doesn't throw!
> runner->decrease_ref();

We protected ourselves from exceptions with the following expression
```
 [runner] () noexcept { runner->wait(); }();
```
Any exception will cause termination due to `noexcept` specifier.

However, we still need to think about better names.

Reply:
See conversation below.